components/languages.js

```diff
@@ -30,30 +30,43 @@ export const LANGUAGES = {
   vn: ['Vietnamese', 'sans-serif']
 };
 
-export const useLanguages = () => {
-  const {
-    query: { lang: languageFromUrl }
-  } = useRouter();
-
-  let currentLanguage = 'gb';
-  if (languageFromUrl && Object.prototype.hasOwnProperty.call(LANGUAGES, languageFromUrl)) {
-    currentLanguage = languageFromUrl;
-  }
-
-  const [name, fontFamily] = LANGUAGES[currentLanguage];
-
-  // eslint-disable-next-line import/no-dynamic-require
-  const BODY = require(`../language/${name}/index.mdx`);
-  // eslint-disable-next-line import/no-dynamic-require
-  const FAQ = require(`../language/${name}/faq.mdx`);
-
-  return [
-    {
-      name: currentLanguage,
-      fontFamily,
-      body: BODY,
-      faq: FAQ
-    },
-    Object.keys(LANGUAGES).sort()
-  ];
+const LanguageContext = React.createContext(null);
+
+export const useLanguages = () => useContext(LanguageContext);
+
+export const withLanguages = (Page) => {
+  const WithLanguages = ({ currentLanguage, flags, ...props }) => (
+    <LanguageContext.Provider value={{ currentLanguage, flags }}>
+      <Page currentLanguage={currentLanguage} flags={flags} {...props} />
+    </LanguageContext.Provider>
+  );
+
+  WithLanguages.getInitialProps = async (context) => {
+    const languageFromUrl = context.query.lang;
+
+    const currentLanguage =
+      languageFromUrl && Object.prototype.hasOwnProperty.call(LANGUAGES, languageFromUrl)
+        ? languageFromUrl
+        : 'gb';
+
+    const [name, fontFamily] = LANGUAGES[currentLanguage];
+
+    const BODY = await import(`../language/${name}/index.mdx`);
+    const FAQ = await import(`../language/${name}/faq.mdx`);
+
+    return {
+      ...(Page.getInitialProps ? await Page.getInitialProps(context) : {}),
+      currentLanguage: {
+        name: currentLanguage,
+        fontFamily,
+        body: BODY,
+        faq: FAQ
+      },
+      flags: Object.keys(LANGUAGES).sort()
```

Comment:
Does flags need to be in the context? It's only used in the header and in there it's pulled from the hook.

Reply:
Never mind, I saw the hook utilises the context so it is necessary.